sdk/src/azure/core/az_json_parser.c

```diff
@@ -1,558 +1,693 @@
+
+  _az_json_stack_push(&json_parser->_internal.bit_stack, container_kind);
+  _az_json_parser_update_state(json_parser, token_kind, az_span_slice(token, 0, 1), 1);
   return AZ_OK;
 }
 
-AZ_NODISCARD AZ_INLINE az_result az_json_parser_pop_stack(az_json_parser* json_parser)
+AZ_NODISCARD static bool _az_is_valid_escaped_character(uint8_t byte)
 {
-  return az_json_stack_pop(&json_parser->_internal.stack);
+  switch (byte)
+  {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+    default:
+      return false;
+  }
 }
 
-AZ_NODISCARD az_result az_json_parser_init(az_json_parser* json_parser, az_span json_buffer)
+// A hex digit is valid if it is in the range: [0..9] | [A..F] | [a..f]
+// Otherwise, return false.
+AZ_NODISCARD static bool _az_is_hex_digit(uint8_t byte)
 {
-  *json_parser = (az_json_parser){ ._internal = { .reader = json_buffer, .stack = 1 } };
-  return AZ_OK;
+  return (byte >= '0' && byte <= '9') || (byte >= 'a' && byte <= 'f')
+      || (byte >= 'A' && byte <= 'F');
 }
 
-// 18 decimal digits. 10^18 - 1.
-//                        0         1
-//                        012345678901234567
-#define AZ_DEC_NUMBER_MAX 999999999999999999ull
-
-typedef struct
-{
-  int sign;
-  uint64_t value;
-  bool remainder;
-  int16_t exp;
-} _az_dec_number;
-
-/* Calculate 10 ^ exp with O(log exp) by doing incremental multiplication
- If result goes beyone double limits (Overflow), infinite is returned based on standard IEEE_754
- https://en.wikipedia.org/wiki/IEEE_754.
- If Underflow, 0 is returned
-*/
-AZ_NODISCARD static double _ten_to_exp(int16_t exp)
+AZ_NODISCARD static bool _az_validate_hex_digits(uint8_t* token_ptr, int32_t index)
 {
-  double result = 1;
-  double incrementing_base = 10;
-  int16_t abs_exp = (int16_t)(exp < 0 ? -exp : exp);
-
-  while (abs_exp > 0)
+  // The caller already guaranteed that we have at least 4 bytes in the buffer.
+  for (int32_t i = 0; i < 4; i++)
   {
-    // odd exp would update result to current incremented base
-    if (abs_exp & 1)
+    uint8_t next_byte = token_ptr[index + i];
+
+    if (!_az_is_hex_digit(next_byte))
     {
-      result = exp < 0 ? result / incrementing_base : result * incrementing_base;
+      return false;
     }
-
-    abs_exp /= 2;
-    incrementing_base = incrementing_base * incrementing_base;
   }
-  return result;
+  return true;
 }
 
-// double result follows IEEE_754 https://en.wikipedia.org/wiki/IEEE_754
-static AZ_NODISCARD az_result _az_json_number_to_double(_az_dec_number const* p, double* out)
+AZ_NODISCARD static az_result _az_json_parser_process_string(az_json_parser* json_parser)
 {
-  *out = (double)p->value * _ten_to_exp(p->exp) * (double)p->sign;
-  return AZ_OK;
-}
+  // Move past the first '"' character
+  json_parser->_internal.bytes_consumed++;
+
+  az_span token = _get_remaining_json(json_parser);
+  uint8_t* const token_ptr = az_span_ptr(token);
```

Comment:
@antkmsft, there were a bunch of parameter renaming conflicts in this file based on https://github.com/Azure/azure-sdk-for-c/pull/819. I kept most of the const changes where possible, but none of the `self to ref_span` or `p_reader to  reader` renames apply anymore.

Also, I didn't fully understand why you changed some `json_parser` parameters to `ref_json_parser`, within the helper methods. The public surface area wasn't changed, so I kept things as `json_parser`. What was the rationale for only applying that parameter rename, selectively?

Please take a look to make sure I didn't accidentally undo some of your changes that I should add back.

Reply:
It is easy to remember, same as in C#: `ref_` is for in/out parameters. `out_` is for out parameters. If it is read only, even though it is a pointer, no prefix is needed,

Reply:
And normally, if a parameter is `T*`, it is at the very least `ref_`, and sometimes `out_`, because a function can write to the location it points to.
If it is `T const*`, I don't see how it can be `ref_` or `out_`.

Reply:
Please try not to undo my renaming changes where applicable, because I've correctly addressed 99% of the namings, put tome in it, and unified it :)

Reply:
> Please try not to undo my renaming changes where applicable, because I've correctly addressed 99% of the namings, put tome in it, and unified it :)

I did not. The ones in your previous PR that affects `az_json_parse`r, no longer apply to this PR anymore since the code changed substantially. I will revisit this again in the future, to make sure the parameter naming rules you described are followed.

https://github.com/Azure/azure-sdk-for-c/pull/819/files#diff-36e2a004054a75fbdbed6861091ae0d3R46